src/main/java/st/ratpack/auth/CachingTokenValidator.java

```diff
@@ -22,38 +22,21 @@ public CachingTokenValidator(TokenValidator upstreamValidator) {
 			.maximumSize(10000L)
 			.expireAfterWrite(5L, TimeUnit.MINUTES)
 			//			.executor(Execution.current().getEventLoop())  Don't do this it makes Ratpack hang.
-			.build();
+			.build(this::loadCache);
 	}
 
 	@Override
 	public Promise<Optional<OAuthToken>> validate(String token) {
-		Promise<Optional<OAuthToken>> promiseOAuthToken = cache.getIfPresent(token);
-
-		if (promiseOAuthToken == null) {
-			//Cache miss use the upstream validator
-			logger.trace("Cache MISS: " + token);
-			promiseOAuthToken = upstreamValidator.validate(token);
-
-			//Make sure we only call the validate on the upstream once
-			promiseOAuthToken = promiseOAuthToken.cache();
-
-			promiseOAuthToken.then(optionalOAuth -> {
-				logger.trace("PUTTING: " + optionalOAuth);
-				cache.put(token, Promise.value(optionalOAuth));
```

Comment:
[AsyncLoadingCache](https://github.com/ben-manes/caffeine/wiki/Population#asynchronously-loading) will invalidate the entry if the future fails. This is done by adding a `whenComplete` handler so that it is processed right after the future completes. This is because caches do not support _negative caching_ natively and leave that as a simple user-level optimization, which is seems to be the least error prone / surprising approach.

Ratpack has its own concurrency model so perhaps using `AsyncLoadingCache` and adapting a `CompletableFuture` to a `Promise` isn't ideal. It may be better to have the loader function return a Promise that calls `cache.invalidate(key)` in an `onError` block.